pipewire/src/main_loop.rs

```diff
@@ -0,0 +1,449 @@
+use pipewire_native_spa as spa;
+use spa::dict::Dict;
+use spa::flags;
+use spa::interface;
+use spa::interface::cpu::CpuImpl;
+use spa::interface::ffi::{CControlHooks, CHook};
+use spa::interface::log::{LogImpl, LogLevel};
+use spa::interface::r#loop::{
+    LoopControlMethodsImpl, LoopImpl, LoopUtilsImpl, LoopUtilsSource, SourceEventFn, SourceIdleFn,
+    SourceIoFn, SourceSignalFn, SourceTimerFn,
+};
+use spa::interface::system::SystemImpl;
+use spa::support::ffi;
+use spa::{emit_hook, hook::HookList};
+
+use std::cell::RefCell;
+use std::os::fd::RawFd;
+use std::path::PathBuf;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::sync::mpsc::{channel, Receiver, Sender};
+
+pub struct MainLoopEvents {
+    destroy: Box<dyn FnMut()>,
+}
+
+struct Loop {
+    #[allow(dead_code)]
+    system: Rc<Pin<Box<SystemImpl>>>,
+    r#loop: Rc<Pin<Box<LoopImpl>>>,
+    control: Rc<Pin<Box<LoopControlMethodsImpl>>>,
+    utils: Rc<Pin<Box<LoopUtilsImpl>>>,
+    name: String,
+}
+
+#[allow(dead_code)]
+pub struct MainLoop {
+    log_handle: Box<dyn interface::plugin::Handle>,
+    system_handle: Box<dyn interface::plugin::Handle>,
+    cpu_handle: Box<dyn interface::plugin::Handle>,
+    loop_handle: Box<dyn interface::plugin::Handle>,
+    support: interface::Support,
+    plugin: ffi::plugin::Plugin,
+
+    pw_loop: Loop,
+    hooks: Rc<RefCell<HookList<MainLoopEvents>>>,
+    sender: Option<Sender<()>>,
+    receiver: Option<Receiver<()>>,
```

Comment:
Any reason to prefer a sender/receiver here to something like what the C code does (i.e. a boolean flag that is modified via `invoke()` to stop the loop)?